File: Interview Problems/Parking Lot/Java Code/repository/TicketRepository.java
package repository;

import domain.Ticket;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.stream.Collectors;

public class TicketRepository {

    private final Map<UUID, Ticket> store = new HashMap<>();

    public void save(Ticket ticket) {
        store.put(ticket.getId(), ticket);
        System.out.println("[REPO] Ticket saved: " + ticket.getId());
    }

    public Optional<Ticket> findById(UUID ticketId) {
        return Optional.ofNullable(store.get(ticketId));
    }

    public List<Ticket> findActiveTickets() {
        return store.values().stream()
                .filter(Ticket::isActive)
                .collect(Collectors.toList());
    }

    /**
     * Deactivates a ticket to prevent it being reused for a second exit.
     */
    public void deactivateTicket(UUID ticketId) {
        Ticket ticket = store.get(ticketId);
        if (ticket == null) throw new IllegalArgumentException("Ticket not found: " + ticketId);
        ticket.deactivate();
        System.out.println("[REPO] Ticket deactivated: " + ticketId);
    }

    public void update(Ticket ticket) {
        if (!store.containsKey(ticket.getId())) {
            throw new IllegalArgumentException("Ticket not found: " + ticket.getId());
        }
        store.put(ticket.getId(), ticket);
    }
}

File: Interview Problems/Parking Lot/Python Code/repository/ticket_repository.py
import uuid
from typing import Dict, List, Optional
from domain.ticket import Ticket

class TicketRepository:
    def __init__(self):
        self._store: Dict[uuid.UUID, Ticket] = {}

    def save(self, ticket: Ticket):
        self._store[ticket.id] = ticket
        print(f"[REPO] Ticket saved: {ticket.id}")

    def find_by_id(self, ticket_id: uuid.UUID) -> Optional[Ticket]:
        return self._store.get(ticket_id)

    def find_active_tickets(self) -> List[Ticket]:
        return [t for t in self._store.values() if t.is_active]

    def deactivate_ticket(self, ticket_id: uuid.UUID):
        ticket = self._store.get(ticket_id)
        if not ticket:
            raise ValueError(f"Ticket not found: {ticket_id}")
        ticket.deactivate()
        print(f"[REPO] Ticket deactivated: {ticket_id}")

    def update(self, ticket: Ticket):
        if ticket.id not in self._store:
            raise ValueError(f"Ticket not found: {ticket.id}")
        self._store[ticket.id] = ticket
